### PR TITLE
[FSDP2] Added test to show rank 0 CPU full state dict flow

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124787
* #124768
* #124767
* __->__ #124741
* #124651

This PR adds a unit test to show how we can convert FSDP2 GPU sharded state dicts to a CPU full state dict on rank 0.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: test/distributed/_composable/fsdp/test_fully_shard_state_dict.py [hunk 1]
```diff
@@ -158,5 +165,59 @@ def _test_state_dict_save_load(self, model: nn.Module):
...
+        return 2
+
+    @unittest.skipIf(not TEST_CUDA, "no cuda")
+    def test_rank0_offload_full_state_dict(self):
+        # Construct a reference unsharded model on all ranks
+        model_args = ModelArgs(dropout_p=0.0)
+        torch.manual_seed(42)
+        ref_model = Transformer(model_args).cuda()
+        for param in ref_model.parameters():
+            torch.distributed.broadcast(param.detach(), src=0)
```
Comment (wanchaol) [inline test/distributed/_composable/fsdp/test_fully_shard_state_dict.py:180]: do we have some logic in `fully_shard` to broadcast the initial tensors?

**Review by wanchaol — APPROVED** [review]

File: test/distributed/_composable/fsdp/test_fully_shard_state_dict.py (hunk 1 above)
Comment (awgu) [reply test/distributed/_composable/fsdp/test_fully_shard_state_dict.py:180]: We do not right now. This broadcast here is actually because we are using MTPG, so I think that there is no way to ensure the same starting reference model on all ranks (since seed is per process, not per thread).

**Review by wz337 — APPROVED** [review]
